internal/controller/drpolicy_peerclass.go

```diff
@@ -0,0 +1,590 @@
+	if err != nil {
+		return err
+	}
+
+	return pruneClassViews(m, log, clusterName, survivorClassNames, mcvList)
+}
+
+// getVSClassesFromCluster gets VolumeSnapshotClasses that are claimed in the ManagedClusterInstance
+func getVSClassesFromCluster(
+	u *drpolicyUpdater,
+	m util.ManagedClusterViewGetter,
+	mc *util.ManagedClusterInstance,
+	clusterName string,
+) ([]*snapv1.VolumeSnapshotClass, error) {
+	vsClasses := []*snapv1.VolumeSnapshotClass{}
+
+	vsClassNames := mc.VolumeSnapshotClassClaims()
+	if len(vsClassNames) == 0 {
+		return vsClasses, nil
+	}
+
+	annotations := make(map[string]string)
+	annotations[AllDRPolicyAnnotation] = clusterName
+
+	for _, vscName := range vsClassNames {
+		sClass, err := m.GetVSClassFromManagedCluster(vscName, clusterName, annotations)
+		if err != nil {
+			return []*snapv1.VolumeSnapshotClass{}, err
+		}
+
+		vsClasses = append(vsClasses, sClass)
+	}
+
+	return vsClasses, pruneVSClassViews(m, u.log, clusterName, vsClassNames)
+}
+
+func pruneSClassViews(
+	m util.ManagedClusterViewGetter,
+	log logr.Logger,
+	clusterName string,
+	survivorClassNames []string,
+) error {
+	mcvList, err := m.ListSClassMCVs(clusterName)
+	if err != nil {
+		return err
+	}
+
+	return pruneClassViews(m, log, clusterName, survivorClassNames, mcvList)
+}
+
+// getSClassesFromCluster gets StorageClasses that are claimed in the ManagedClusterInstance
+func getSClassesFromCluster(
+	u *drpolicyUpdater,
+	m util.ManagedClusterViewGetter,
+	mc *util.ManagedClusterInstance,
+	clusterName string,
+) ([]*storagev1.StorageClass, error) {
+	sClasses := []*storagev1.StorageClass{}
+
+	sClassNames := mc.StorageClassClaims()
+	if len(sClassNames) == 0 {
+		return sClasses, nil
+	}
+
+	annotations := make(map[string]string)
+	annotations[AllDRPolicyAnnotation] = clusterName
+
+	for _, scName := range sClassNames {
+		sClass, err := m.GetSClassFromManagedCluster(scName, clusterName, annotations)
+		if err != nil {
+			return []*storagev1.StorageClass{}, err
+		}
+
+		sClasses = append(sClasses, sClass)
+	}
+
+	return sClasses, pruneSClassViews(m, u.log, clusterName, sClassNames)
+}
+
+// getClusterClasses inspects, using ManagedClusterView, the ManagedCluster claims for all storage related classes,
+// and returns the set of classLists for the passed in clusters
+func getClusterClasses(
+	u *drpolicyUpdater,
+	m util.ManagedClusterViewGetter,
+	cluster string,
+) (classLists, error) {
+	mc, err := util.NewManagedClusterInstance(u.ctx, u.client, cluster)
+	if err != nil {
+		return classLists{}, err
+	}
+
+	clID, err := mc.ClusterID()
+	if err != nil {
+		return classLists{}, err
+	}
+
+	sClasses, err := getSClassesFromCluster(u, m, mc, cluster)
+	if err != nil || len(sClasses) == 0 {
+		return classLists{}, err
+	}
+
+	vsClasses, err := getVSClassesFromCluster(u, m, mc, cluster)
+	if err != nil || len(sClasses) == 0 {
+		return classLists{}, err
+	}
+
+	vrClasses, err := getVRClassesFromCluster(u, m, mc, cluster)
```

Comment:
If only VolSync is enabled, VRC is not needed at that point. Isn't this going to always fail the DRPolicy reconciliation?

Reply:
Relooking at this, I think it is a copy paste mistake, will report back.

Reply:
Ok the len check on SCClass list was superfluous and a copy/paste issue (i.e after the initial scClasses len check as != 0 it will pass for other checks below).

On the original review comment:
- If a cluster claims a VRClass, only then do we attempt to read it, so for cluster that does not have VR setups, this will return an empty vrClass list and nil err

- If a cluster claims a VRClass, then we want to be sure to read it and process it, so any failure to read it is treated as an error processing the cluster. This is done so that we do not form an initial peerClass that would default workloads to Volsync and then later update it to Volrep, at which point the workload will not shift replication schemes

Reply:
The len check is fixed.